runtime/userspace/api/caliptra-api/src/firmware_update/mod.rs

```diff
@@ -0,0 +1,294 @@
+    }
+
+    async fn update_caliptra(&mut self, offset: usize, len: usize) -> Result<(), ErrorCode> {
+        let mut req = MailboxReqHeader { chksum: 0 };
+        let req_data = req.as_mut_bytes();
+        self.mailbox
+            .populate_checksum(CommandId::FIRMWARE_LOAD.into(), req_data)
+            .unwrap();
+
+        let response_buffer = &mut [0u8; core::mem::size_of::<GetImageInfoResp>()];
+
+        let mut payload_stream = MailboxPayloadStream::new(self.staging_memory, offset, len);
+
+        loop {
+            let result = self
+                .mailbox
+                .execute_with_payload_stream(
+                    CommandId::FIRMWARE_LOAD.into(),
+                    None,
+                    &mut payload_stream,
+                    response_buffer,
+                )
+                .await;
+            match result {
+                Ok(_) => return Ok(()),
+                Err(MailboxError::ErrorCode(ErrorCode::Busy)) => continue,
+                Err(_) => return Err(ErrorCode::Fail),
+            }
+        }
+    }
+
+    async fn update_manifest(&mut self, offset: usize, len: usize) -> Result<(), ErrorCode> {
+        let mut req = AuthManifestReqHeader {
+            chksum: 0,
+            manifest_size: len as u32,
+        };
+
+        let mut payload_stream = MailboxPayloadStream::new(self.staging_memory, offset, len);
+
+        // Calculate the mailbox checksum
+        let mut checksum = payload_stream.get_bytesum().await;
+        for b in CommandId::SET_AUTH_MANIFEST.0.to_le_bytes().iter() {
+            checksum = checksum.wrapping_add(u32::from(*b));
+        }
+        for b in req.as_mut_bytes().iter() {
+            checksum = checksum.wrapping_add(u32::from(*b));
+        }
+        req.chksum = 0u32.wrapping_sub(checksum);
+
+        let response_buffer = &mut [0u8; core::mem::size_of::<MailboxRespHeader>()];
+        let header = req.as_mut_bytes();
+        loop {
+            let result = self
+                .mailbox
+                .execute_with_payload_stream(
+                    CommandId::SET_AUTH_MANIFEST.into(),
+                    Some(header),
+                    &mut payload_stream,
+                    response_buffer,
+                )
+                .await;
+            match result {
+                Ok(_) => return Ok(()),
+                Err(MailboxError::ErrorCode(ErrorCode::Busy)) => continue,
+                Err(_) => return Err(ErrorCode::Fail),
+            }
+        }
+    }
+
+    async fn wait_caliptra_rt_execution(&mut self) -> Result<(), ErrorCode> {
+        let mut req = MailboxReqHeader { chksum: 0 };
+        let req_data = req.as_mut_bytes();
+        self.mailbox
+            .populate_checksum(CommandId::FW_INFO.into(), req_data)
+            .unwrap();
+
+        let response_buffer = &mut [0u8; core::mem::size_of::<FwInfoResp>()];
+
+        // Wait indefinitely until Caliptra RT is ready
+        // Todo: Implement a timeout mechanism
```

Comment:
Our watchdog timer perhaps. Though if we time out, we'll probably want to signal the SoC to restart everything so we can do the recover flow.